Add provider backed by libcrux

keks opened the pull request:
This draft PR adds a provider backed by libcrux. Note that libcrux is not supported on all platforms at the moment, so the tests do not run everywhere.

Fixes https://github.com/openmls/openmls/issues/1319, https://github.com/openmls/openmls/issues/1250

keks commented:
The failing coverage build looks like [this issue](https://github.com/dalek-cryptography/curve25519-dalek/pull/619) and the spot fuzz fails to resolve `api.github.com`. I've restarted it, maybe it works this time.

kkohbrok commented on libcrux_crypto/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "openmls_libcrux_crypto"
+version = "0.1.0"
+edition = "2021"
```
Could you add some metadata here (authors license, description, etc.) like we have in the other Cargo.tomls?

kkohbrok commented on libcrux_crypto/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
This line can probably go.

kkohbrok commented on libcrux_crypto/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+
+[dependencies]
+getrandom = "0.2.12"
+libcrux = { git = "https://github.com/cryspen/libcrux", features = ["rand"] }
```
Any chance we can pin this on a version?

keks replied:
I pinned it to a commit; we don't have a recent release yet

kkohbrok commented on libcrux_crypto/src/crypto.rs:
```diff
@@ -0,0 +1,578 @@
+use rand::CryptoRng;
+use tls_codec::SecretVLBytes;
+
+const MAX_DATA_LEN: usize = 0x10000000;
```
Where is this number coming from? Could you add a code comment?

keks replied:
I was sure I got that from the called code, but I couldn't find it. Reading the AEAD RFC told me that this number is different for each ciphersuite, so I just removed it and let libcrux take care of it. The errors will be less specific but such is life.

kkohbrok commented on libcrux_crypto/src/crypto.rs:
```diff
@@ -0,0 +1,578 @@
+        info: &[u8],
+        aad: &[u8],
+        ptxt: &[u8],
+    ) -> HpkeCiphertext {
```
Would it make sense to patch the crypto provider trait to return a `Result` here? There are quite a few unwraps in this function.

keks replied:
noticed that the existing provider does the same unwraps, and that it might have been better to do this in a separate PR, but now it's done

kkohbrok commented on libcrux_crypto/src/crypto.rs:
```diff
@@ -0,0 +1,578 @@
+            .generate_vec(libcrux::hpke::kem::Nsk(config.1))
+            .map_err(|_| CryptoError::CryptoLibraryError)?;
+
+        let pk_r = libcrux::hpke::kem::DeserializePublicKey(config.1, pk_r).unwrap();
```
Can't this be turned into a `CryptoError`?

kkohbrok commented on libcrux_crypto/src/crypto.rs:
```diff
@@ -0,0 +1,578 @@
+}
+
+impl Default for CryptoProvider {
+    fn default() -> Self {
```
Is it that important for the provider to implement `Default` given that there are two unwraps here?

keks replied:
Don't we use default all over the tests in the OpenMLS tests? If we don't implement it, I don't think we would be able to run the tests against this provider.

kkohbrok replied:
Fair enough, in which case the unwraps are okay, as long as the implementation is qualified with the right config flags.

kkohbrok commented on libcrux_crypto/src/crypto.rs:
```diff
@@ -0,0 +1,578 @@
+    fn derive_hpke_keypair(&self, config: HpkeConfig, ikm: &[u8]) -> HpkeKeyPair {
+        let config = hpke_config(config);
+        let HPKEConfig(_, alg, _, _) = config;
+        let (sk, pk) = hpke::kem::DeriveKeyPair(alg, ikm).unwrap(); // XXX return err
```
☝🏻

kkohbrok reviewed (changes requested):
Thanks for the PR! Generally looks good to me. Some comments inline and two observations:

- There is quite a lot of unwrapping going on that we are generally trying to avoid. I commented on some instances. I'm not sure how to handle the mutex access, but it might make sense to have an error for that as well.
- Libcrux doesn't support all OSs and architectures and some ciphersuites only if AES-NI is available. That's all fine, but we should probably document that somewhere. Maybe on the `CryptoProvider` struct? Or a page in the book?

keks commented:
Thanks! Regarding your two points: I removed all the unwraps except the ones in `default` and replaced them with errors. I also added a section in the book that mentions the limitations.

kkohbrok reviewed (approved):
Thanks for addressing my concerns! Let's get this in and then deal with the (potentially) panicking`Default` implementations in a follow-up.
